### PR TITLE
i18n: Add support for joining XML localization via itstool

### PR DESCRIPTION
**Pull request by ximion**
Hi!

This PR adds support for the "join" feature of `itstool` to Meson. While (x)gettext can merge XML, when working with AppStream metadata I found out that it is pretty unpredictable in how data is merged, especially when inline tags are involved. At the very least, inline markup gets XML-escaped, and I found no way at all to work around this particular issue.
[Itstool](http://itstool.org/) however always generates correct results and obeys the rules defined in ITS files properly. This tool is also very widespread and available in every Linux distribution I checked. It also is used by GNOME's help.

Using `itstool` from Meson is pretty cumbersome at the moment, as you need to scrape through the output generated from `i18n.gettext()` and rename mo files so itstool can infer locale names.
This is admittedly a bit dumb, but there's also no real way around this (even if itstool gained support to specify per-file locale on the command line, it would take ages before the changes are widespread enough to be used unconditionally).
The new code comes with a pretty realistic example of how this feature will likely be used in Meson as unit test.

If `itstool_join` is called we fail if the tool could not be found (as opposed to `merge_file`). I chose this behavior because a silent failure would mean crucial data may not get installed, and even if it was installed would be untranslated which explicitly was not the intent of the developer when the join command is used.
Besides that, I don't think there's anything else particularly noteworthy about this change.

Thanks for considering!

### CODE REVIEW

**Comment by ximion** [comment]
Looks like on bionic, itstool does not generate the expected output - I'll try to reproduce that locally to see what's going on there (and either fix the issue or add a smarter test).

**Comment by ximion** [comment]
That bionic build failure was caused by an older itstool version which did not read custom ITS files correctly while joining. I adjusted the test to pass even if we have an older itstool which has that bug.

File: test cases/frameworks/6 gettext/data3/meson.build [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+if itstool.found()
+
+
+mi_translated = i18n.itstool_join(
+  input:  'com.mesonbuild.test.intlprog.metainfo.xml',
+  output: 'com.mesonbuild.test.intlprog.metainfo.xml',
+  mo_targets: mo_targets,
+  its_files: ['metainfo.its'],
+  install: true,
+  install_dir: join_paths(get_option('datadir'), 'metainfo'),
```
Comment (jpakkane) [inline test cases/frameworks/6 gettext/data3/meson.build:11]: This is old-style code, the modern equivalent should be `get_option('datadir') / 'metainfo'`.

File: test cases/frameworks/6 gettext/data3/meson.build [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+)
+
+# older versions of itstool have a bug where ITS rules specified on the command-line
+# are not read when joining files. Since we don't install appstream in the Meson CI
+# environment, the to-be-tested entry will be untranslated and the test would fail, so
+# we just skip verification if the installed itstool is too old.
+r = run_command(itstool, '-v', check: true)
+itstool_v = r.stdout().strip().split()
+if itstool_v[1].version_compare('>=2.0.6')
+  grep_exe = find_program('grep')
```
Comment (jpakkane) [inline test cases/frameworks/6 gettext/data3/meson.build:21]: This is _not_ acceptable. Any use of shell tools is absolutely prohibited as all tests must run on a bog standard Windows `cmd.exe` without any unixy stuff. This should be a standalone Python script that reads the result file and validates it.

File: test cases/frameworks/6 gettext/data3/meson.build [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            mi_translated]
+  )
+else
+  message('Skipping translation verification: Itstool too old.')
+endif
+
+
+else
+install_data('com.mesonbuild.test.intlprog.metainfo.xml',
+    install_dir: join_paths(get_option('datadir'), 'metainfo'))
```
Comment (jpakkane) [inline test cases/frameworks/6 gettext/data3/meson.build:34]: This should use `/` as well.

**Review by jpakkane — CHANGES_REQUESTED** [review]

**Comment by eli-schwartz** [comment]
> Using `itstool` from Meson is pretty cumbersome at the moment, as you need to scrape through the output generated from `i18n.gettext()` and rename mo files so itstool can infer locale names.
This is admittedly a bit dumb, but there's also no real way around this (even if itstool gained support to specify per-file locale on the command line, it would take ages before the changes are widespread enough to be used unconditionally).

It will take even longer if no one ever pushes to have this changed, so it never gets implemented. If we are going to handle this I'd vastly, vastly prefer to have two codepaths, and choose which one to use depending on the detected version of itstool:
- one where this is done properly
- one where strange file copying happens invisibly via private scripts that hide itstool from ninja and circumvent find_program_impl and repeatedly create tempdirs to repeatedly do file IO

**Comment by eli-schwartz** [comment]
Hmmm... When I was fixing shared-mime-info use of itstool that ran a custom shellscriot which invoked ninja to build mo files, my first implementation actually ran cp to copy those files as ninja build rules, and then just ran itstool directly from ninja. This is more portably done as a meson internal copy.py script... Honestly I like that thought a lot better than layering it in a script with a tempdir.

However I never committed that code because I realized it could just use the better project via i18.merge_file; maybe gettext itself could be fixed to not have the bugs you say are there?

It is a bit inconvenient that itstool requires running msgfmt to generate .mo files which are never installed, when gettext can directly merge .po files into your xml.

**Comment by ximion** [comment]
@eli-schwartz 
> It will take even longer if no one ever pushes to have this changed, so it never gets implemented. If we are going to handle this I'd vastly, vastly prefer to have two codepaths, and choose which one to use depending on the detected version of itstool

Sure, if/when `itstool` has gained support for this, we can do it! Annoyingly, a mo-file does not contain the locale string that it belongs to, so there is nothing clever we can do here, so I am not even sure how a proper solution that does not depend on file names would look like (it would always be some kind of heuristic for semi-clever locale name autodetection).

> It is a bit inconvenient that itstool requires running msgfmt to generate .mo files which are never installed, when gettext can directly merge .po files into your xml.

True, both tools are surprisingly inconvenient in respective corner cases... But at least `itstool` with its more limited scope does the job properly...

@jpakkane I addressed your style comments, and the resulting data is also now verified by a small Python script instead of using `grep` & Co.

File: test cases/frameworks/6 gettext/data3/meson.build [hunk 4]
```diff
@@ -0,0 +1,36 @@
+
+if itstool.found()
+
+
+mi_translated = i18n.itstool_join(
```
Comment (jpakkane) [inline test cases/frameworks/6 gettext/data3/meson.build:5]: These should be indented.

Comment (ximion) [reply test cases/frameworks/6 gettext/data3/meson.build]: Done :-)

**Comment by eli-schwartz** [comment]
> Sure, if/when `itstool` has gained support for this, we can do it!

Are you saying you've submitted a bug report to itstool for this? Great. Please link it here though.

**Comment by ximion** [comment]
> > Sure, if/when `itstool` has gained support for this, we can do it!
> 
> Are you saying you've submitted a bug report to itstool for this? Great. Please link it here though.

There is https://github.com/itstool/itstool/issues/32 and https://github.com/itstool/itstool/issues/44 on this issue (with the latter being the specific issue that we work around here).

File: docs/markdown/i18n-module.md [hunk 5]
```diff
@@ -56,3 +56,15 @@ for normal keywords. In addition it accepts these keywords:
...
+
+This joins translations into a XML file using `itstool`. See
+[[@custom_tgt]]
+for normal keywords. In addition it accepts these keywords:
+
+* `its_files`: filenames of ITS files that should be used explicitly
+  (XML translation rules are autodetected otherwise).
+* `mo_targets` *required*: mo file generation targets as returned by `i18n.gettext()`.
+
+*Added 0.61.0*
```
Comment (eli-schwartz) [inline docs/markdown/i18n-module.md:70]: This is out of date.

File: mesonbuild/modules/i18n.py [hunk 6]
```diff
@@ -269,5 +288,65 @@ def gettext(self, state: 'ModuleState', args: T.Tuple[str], kwargs: 'Gettext') -
 
         return ModuleReturnValue([gmotargets, pottarget, updatepotarget], targets)
 
+    @FeatureNew('i18n.itstool_join', '0.61.0')
```
Comment (eli-schwartz) [inline mesonbuild/modules/i18n.py:291]: This is out of date.

File: mesonbuild/modules/i18n.py [hunk 7]
```diff
@@ -269,5 +288,65 @@ def gettext(self, state: 'ModuleState', args: T.Tuple[str], kwargs: 'Gettext') -
...
+        CT_INSTALL_TAG_KW,
+        CT_OUTPUT_KW,
+        INSTALL_KW,
+        _ARGS.evolve(),
+        KwargInfo('its_files', ContainerTypeInfo(list, str)),
+        KwargInfo('mo_targets', ContainerTypeInfo(list, build.CustomTarget), required=True),
+    )
+    def itstool_join(self, state: 'ModuleState', args: T.List['TYPE_var'], kwargs: 'ItsJoinFile') -> ModuleReturnValue:
+        if not shutil.which('itstool'):
+            self.noitstool_error()
```
Comment (eli-schwartz) [inline mesonbuild/modules/i18n.py:307]: This seems like a weird thing to delegate to a dispatch function as it is only used once, and we should really just use the find_program implementation, which respects tools defined in cross files, then pass it as an argument to the helper script.

File: docs/markdown/i18n-module.md (hunk 5 above)
Comment (eli-schwartz) [reply docs/markdown/i18n-module.md:70]: #9867

**Pull request merged**